src/engraving/libmscore/slur.cpp

```diff
@@ -349,139 +366,327 @@ void SlurSegment::adjustEndpoints()
+                PointF clearancePoint2 = clearanceBezier.pointAtPercent(double(i + 1) / double(npoints));
+                clearancePoint1 = toSystemCoordinates.map(clearancePoint1);
+                clearancePoint2 = toSystemCoordinates.map(clearancePoint2);
+                slurRects.push_back(RectF(clearancePoint1, clearancePoint2));
+            }
+            // Check collisions
+            for (Segment* seg = startSeg;;) {
+                if (seg->next() && !seg->next()->isEndBarLineType()) {
+                    seg = seg->next();
+                } else if (seg->measure()->next() && seg->measure()->next()->isMeasure()) {
+                    seg = toMeasure(seg->measure()->next())->first();
+                } else {
+                    break;
+                }
+                Shape segShape = seg->staffShape(startCR->vStaffIdx()).translated(seg->pos() + seg->measure()->pos());
+                // If cross-staff, also add the shape of second staff
+                if (isCrossStaff) {
+                    SysStaff* startStaff = system()->staves().at(startCR->vStaffIdx());
+                    SysStaff* endStaff = system()->staves().at(endCR->vStaffIdx());
+                    double dist = endStaff->y() - startStaff->y();
+                    Shape secondStaffShape;
+                    secondStaffShape.add(seg->staffShape(endCR->vStaffIdx()).translated(seg->pos() + seg->measure()->pos()));
+                    secondStaffShape.translate(PointF(0.0, dist));
+                    segShape.add(secondStaffShape);
+                }
+                for (unsigned i=0; i < slurRects.size(); i++) {
+                    bool leftSection = i < slurRects.size() / 3;
+                    bool midSection = i >= slurRects.size() / 3 && i < 2 * slurRects.size() / 3;
+                    bool rightSection = i >= 2 * slurRects.size() / 3;
+                    if ((leftSection && collision.left)
+                        || (midSection && collision.mid)
+                        || (rightSection && collision.right)) { // If a collision is already found in this section, no need to check again
+                        continue;
+                    }
+                    bool intersection = slur()->up() ? !Shape(slurRects[i]).clearsVertically(segShape)
+                                        : !segShape.clearsVertically(slurRects[i]);
+                    if (intersection) {
+                        if (leftSection) {
+                            collision.left = true;
+                        }
+                        if (midSection) {
+                            collision.mid = true;
+                        }
+                        if (rightSection) {
+                            collision.right = true;
+                        }
+                    }
+                }
+                if (seg == endSeg) {
+                    break;
+                }
+            }
+            // In the even iterations, adjust the shape
```

Comment:
not quite sure about this one. What was the motivation?

Reply:
Basically, if I see that the slur collides with something, I adjust it in small steps: one step, I adjust the shape; next step, I adjust the end points; next I adjust the shape again; next the end points again; until there are no collisions anymore. The simplest way I could think for doing this is to do the shape in the even iterations, and the end points in the odd iterations.